qemu-psa-sp.mk

```diff
@@ -0,0 +1,141 @@
+endef
+
+ifeq ($(SP_PACKAGING_METHOD),fip)
+$(eval $(call add-dtc-define,SPMC_TESTS))
+$(eval $(call add-dtc-define,TS_SMM_GATEWAY))
+
+TF_A_EXPORTS += DTC_CPPFLAGS="$(DTC_CPPFLAGS)"
+endif
+
+OPTEE_OS_COMMON_EXTRA_FLAGS += \
+	CFG_SECURE_PARTITION=y \
+	CFG_CORE_SEL1_SPMC=y \
+	CFG_CORE_HEAP_SIZE=131072 \
+	CFG_DT=y \
+	CFG_MAP_EXT_DT_SECURE=y
+
+SP_SMM_GATEWAY_EXTRA_FLAGS +=-DMM_COMM_BUFFER_ADDRESS="0x00000000 0x42000000"
+TS_APP_COMMON_FLAGS +=-DMM_COMM_BUFFER_ADDRESS="0x42000000"
+
+# The boot order of the SPs is determined by the order of calls here. This is
+# due to the SPMC not (yet) supporting the boot order field of the SP manifest.
+ifneq ($(SPMC_TESTS),y)
+# PSA SPs
+$(eval $(call build-sp,block-storage,config/$(SP_BLOCK_STORAGE_CONFIG),63646e80-eb52-462f-ac4f-8cdf3987519c,$(SP_BLOCK_STORAGE_EXTRA_FLAGS)))
+$(eval $(call build-sp,internal-trusted-storage,config/$(SP_PSA_ITS_CONFIG),dc1eef48-b17a-4ccf-ac8b-dfcff7711b14,$(SP_PSA_ITS_EXTRA_FLAGS)))
+$(eval $(call build-sp,protected-storage,config/$(SP_PSA_PS_CONFIG),751bf801-3dde-4768-a514-0f10aeed1790,$(SP_PSA_PS_EXTRA_FLAGS)))
+$(eval $(call build-sp,crypto,config/$(SP_PSA_CRYPTO_CONFIG),d9df52d5-16a2-4bb2-9aa4-d26d3b84e8c0,$(SP_PSA_CRYPTO_EXTRA_FLAGS)))
+ifeq ($(MEASURED_BOOT),y)
+$(eval $(call build-sp,attestation,config/$(SP_PSA_ATTESTATION_CONFIG),a1baf155-8876-4695-8f7c-54955e8db974,$(SP_PSA_ATTESTATION_EXTRA_FLAGS)))
+endif
+ifeq ($(TS_SMM_GATEWAY),y)
+$(eval $(call build-sp,smm-gateway,config/$(SP_SMM_GATEWAY_CONFIG),ed32d533-99e6-4209-9cc0-2d72cdd998a7,$(SP_SMM_GATEWAY_EXTRA_FLAGS)))
+endif
+else
+# SPMC test SPs
+OPTEE_OS_COMMON_EXTRA_FLAGS	+= CFG_SPMC_TESTS=y
+$(eval $(call build-sp,spm-test1,opteesp,5c9edbc3-7b3a-4367-9f83-7c191ae86a37,$(SP_SPMC_TEST_EXTRA_FLAGS)))
+$(eval $(call build-sp,spm-test2,opteesp,7817164c-c40c-4d1a-867a-9bb2278cf41a,$(SP_SPMC_TEST_EXTRA_FLAGS)))
+$(eval $(call build-sp,spm-test3,opteesp,23eb0100-e32a-4497-9052-2f11e584afa6,$(SP_SPMC_TEST_EXTRA_FLAGS)))
+$(eval $(call build-sp,spm-test4,opteesp,423762ed-7772-406f-99d8-0c27da0abbf8,$(SP_SPMC_TEST_EXTRA_FLAGS)))
+endif
+
+# Linux user space applications
+ifneq ($(SPMC_TESTS),y)
+$(eval $(call build-ts-app,libts))
+$(eval $(call build-ts-app,ts-service-test))
+$(eval $(call build-ts-app,psa-api-test/internal_trusted_storage))
+$(eval $(call build-ts-app,psa-api-test/protected_storage))
+$(eval $(call build-ts-app,psa-api-test/crypto))
+ifeq ($(MEASURED_BOOT),y)
+$(eval $(call build-ts-app,psa-api-test/initial_attestation))
+endif
+ifeq ($(TS_UEFI_TESTS),y)
+$(eval $(call build-ts-app,uefi-test))
+
+# uefi-test uses MM Communicate via the arm-ffa-user driver and the message
+# payload is forwarded in a carveout memory area. Adding reserved-memory node to
+# the device tree to prevent Linux from using the carveout area for other
+# purposes.
+
+CARVEOUT_ENTRY = $(ROOT)/build/qemu_v8/mm_communicate_carveout.dtsi
+$(QEMU_DTB_PATH): $(CARVEOUT_ENTRY) dumpdtb | linux
```

Comment:
I think `dumptdb` should be replaced by the actual file path (here and in place of the `dumptdb:` target in `qemu_v8.mk`), otherwise make will rebuild the same thing each time it is invoked although nothing changed. You may still keep a `dumptdb: path/to/dtb` target for convenience.

Reply:
The reason I want it to be always regenerated is that the arguments (`QEMU_ARGS`) passed to QEMU will have an effect on the generated device tree.

Reply:
Ah, makes sense. Ideally you could generate a file with the value of $(QEMU_ARGS) which would change only when those args change (via some clever make macro), then use that as the true dependency. But that's perhaps for another PR (see for instance check-conf-h and mv-if-changed in the OP-TEE OS makefiles).
Could you at least add a comment here? "# Note: the DTB generated by dumpdtb depends on $(QEMU_ARGS)" or something like that. Thanks!

Reply:
Done. Also created a [commit](https://github.com/stormset/build/commit/58e71d3409835b45347ff18ed820752364260ed1) using the mentioned macros. Will create a PR after this gets merged.

Reply:
> Done. Also created a [commit](https://github.com/stormset/build/commit/58e71d3409835b45347ff18ed820752364260ed1) using the mentioned macros. Will create a PR after this gets merged.

Excellent!

Reply:
https://github.com/OP-TEE/build/pull/695 also generates a DTB for Qemu. Maybe there are some factorization to prepare.